### PR TITLE
[db/server] Link workspaces and projects

### PR DESCRIPTION
**Pull request by AlexTugarev**
We're now linking projects to workspaces as well. 

Also in this PR:
* adding index on `d_b_project.cloneUrl`
* removing `Without` type (replaceable with `Omit`)
* refactored `ProjectDB.storeProject` and `Project.create`

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4846#" title="Author self-approved">AlexTugarev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (jankeromnes) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:29]:
```suggestion
        return repo.findOne({ cloneUrl });
```

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (jankeromnes) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:44]:
```suggestion
        if (repo.findOne({ cloneUrl })) {
```

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (jankeromnes) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:52]:
```suggestion
            cloneUrl,
```

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:44]: Will always be `true`. Missing `await`.

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/project-db-impl.ts:49]: Personally I like to decouple such creation logic from the DB, where it is hard to test etc. One example is to move this into a method `Project.new` in `project-protocol.ts` which takes a `Without<Project, 'id' | 'creationTime'>`.

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (geropl) [inline components/server/ee/src/workspace/workspace-factory.ts:169]: Really? So there is no sharing of prebuilds between teams? Is that enforced somewhere?

File: components/server/src/workspace/workspace-factory.ts
Comment (geropl) [inline components/server/src/workspace/workspace-factory.ts:141]: This reads like a project owns a repository exclusively. Is that meant to be? This feels like it's going to be a 100% shadow of a GitHub org, and I thought this would not cover all our usecases... but maybe have some reading to do, here.

**Comment by geropl** [comment]
The first thing strikes me a bit odd: A workspaces I always `1:0..1` to a project? :thinking: Without thinking deeply about it I always thought about it as `1:0...n` - but would need to think about concrete usecases first.

@jankeromnes Left some remarks. I'm not sure I'm qualified to 'lgtm' before I did some more reading on "Teams & Projects" . :slightly_smiling_face:

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (svenefftinge) [reply components/server/ee/src/workspace/workspace-factory.ts:169]: The current logic is that we have a single project per repo in Gitpod and workspaces are assigned to a project no matter you are a member or not. Happy to discuss 📞  😄

File: components/server/src/workspace/workspace-factory.ts
Comment (svenefftinge) [reply components/server/src/workspace/workspace-factory.ts:141]: Yes, that is how it works. Having multiple projects per repo isn't really supported by GitHub, because I can install the GitHub app only once per org. But also the use cases I've heard so far have been are rather edge cases.

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (AlexTugarev) [reply components/server/ee/src/workspace/workspace-factory.ts:169]: @geropl this association does not reflect the runtime behavior. finding and using a prebuild to start a workspace from may follow other rules.

File: components/gitpod-db/src/typeorm/project-db-impl.ts
Comment (AlexTugarev) [reply components/gitpod-db/src/typeorm/project-db-impl.ts:49]: you're talking about externalizing the creation of `Project` object to be stored afterwards here. 
then we can if `id` included and add the check. 
that sounds good.

File: components/server/src/workspace/workspace-factory.ts
Comment (AlexTugarev) [reply components/server/src/workspace/workspace-factory.ts:141]: > This reads like a project owns a repository exclusively.

A team owns a projects, yes.

A repository is not meant to be used in multiple projects. The `cloneUrl` should be unique.

> this would not cover all our usecases

What do we miss?

File: components/server/ee/src/workspace/workspace-factory.ts
Comment (geropl) [reply components/server/ee/src/workspace/workspace-factory.ts:169]: Discussed offline and I now understand that:
 - project is owned by exactly one team
 - project is a GH repo
 - thus there is a 1-1 relation between workspace and project (atm, at least.

File: components/server/src/workspace/workspace-factory.ts
Comment (geropl) [reply components/server/src/workspace/workspace-factory.ts:141]: See comment above

> What do we miss?

Before I thought about a "Team" as a way to have my custom "config overlay" over arbitrary projects, which are _my_ representation of a GH repo. But I understand that this is not easily doable with GH because of the "write access constraint" necessary for GH Apps, and might be way to complicated for most of our usecases. :+1:

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4846#" title="Author self-approved">AlexTugarev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev]
- ~~[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)~~ [AlexTugarev]
- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by AlexTugarev** [comment]
@jankeromnes, please have another pass on this.

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (jankeromnes) [inline components/gitpod-protocol/src/teams-projects-protocol.ts:38]: `projectId` would be more correct, but this is unimportant

**Review by jankeromnes — APPROVED** [review]
Looks good to me 😊 thanks!

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 10962e03610ebe1d13d7b24d55c3e53ffb3711e6</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4846#" title="Author self-approved">AlexTugarev</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4846#pullrequestreview-710324452" title="Approved">jankeromnes</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev,jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev,jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
